tools/offline_log_viewer/kafka.py

```diff
@@ -32,18 +42,98 @@ def decode(self):
         for batch in self.batches():
             header = batch.header_dict()
             yield header
-            if not self.headers_only:
-                records = []
-                for record in batch:
-                    attrs = header["expanded_attrs"]
-                    is_tx_ctrl = attrs["transactional"] and attrs[
-                        "control_batch"]
-                    record_dict = record.kv_dict()
-                    records.append(record_dict)
-                    if is_tx_ctrl:
-                        record_dict["type"] = self.get_control_record_type(
-                            record.key)
-                    yield record_dict
+            if self.headers_only:
+                continue
+            records = batch
+            if header["type_name"] == "tm_update":
+                records = list(batch)
+                assert len(records) == 1
```

Comment:
is the only reason you are materializing `list(batch)` here so that you can check its length?

if so, consider adding `def __len__()` method to the `Batch` type since it already supports materialization and the number of records is stored there so you can use `len(batch) == 1` without materialization and the interface is easier to use.